ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/assertions/CommandDeduplicationAssertions.scala

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.testtool.infrastructure.assertions
+
+import java.time.{Duration, Instant}
+
+import com.daml.api.util.DurationConversion
+import com.daml.ledger.api.testtool.infrastructure.Assertions.{assertDefined, fail}
+import com.daml.ledger.api.testtool.infrastructure.participant.{
+  CompletionResponse,
+  ParticipantTestContext,
+}
+import com.daml.ledger.api.v1.completion.Completion
+import com.daml.ledger.api.v1.experimental_features.CommandDeduplicationPeriodSupport.{
+  DurationSupport,
+  OffsetSupport,
+}
+import com.daml.ledger.api.v1.ledger_offset.LedgerOffset
+import com.daml.ledger.client.binding.Primitive.Party
+import com.google.protobuf.duration.{Duration => DurationProto}
+import io.grpc.Status.Code
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.math.Ordering.Implicits._
+
+object CommandDeduplicationAssertions {
+
+  def assertDeduplicationDuration(
+      requestedDeduplicationDuration: DurationProto,
+      previousSubmissionSendTime: Instant,
+      completionReceiveTime: Instant,
+      completion: Completion,
+      submittingParty: Party,
+      ledger: ParticipantTestContext,
+  )(implicit executionContext: ExecutionContext): Future[Unit] = {
+    val requestedDuration = DurationConversion.fromProto(requestedDeduplicationDuration)
+    ledger.features.commandDeduplicationFeatures.getDeduplicationPeriodSupport.durationSupport match {
+      case DurationSupport.DURATION_NATIVE_SUPPORT =>
+        val reportedDurationProto = assertDefined(
+          completion.deduplicationPeriod.deduplicationDuration,
+          "No deduplication duration has been reported",
+        )
+        val reportedDuration = DurationConversion.fromProto(reportedDurationProto)
+        assert(
+          reportedDuration >= requestedDuration,
+          s"The reported deduplication duration $reportedDuration was smaller than the requested deduplication duration $requestedDuration.",
+        )
+        Future.unit
+      case DurationSupport.DURATION_CONVERT_TO_OFFSET =>
+        val reportedOffset = assertDefined(
+          completion.deduplicationPeriod.deduplicationOffset,
+          "No deduplication offset has been reported",
+        )
+        if (completion.getStatus.code == Code.ALREADY_EXISTS.value) {
+          val completionStreamRequest = ledger.completionStreamRequest(
+            LedgerOffset.of(LedgerOffset.Value.Absolute(reportedOffset))
+          )(submittingParty)
+          ledger
+            .findCompletion(completionStreamRequest)(_.commandId == completion.commandId)
```

Comment:
I'm not too familiar with what `findCompletion` does, but this looks as if you're trying to find any completion for the given command ID. You should restrict the filter to accepting completions (because only those trigger deduplication) and you should make sure that you stop looking for completions before the `completion` (otherwise you might find another later completion for the same command ID; probably not for the current test cases, but maybe in the future?).

Reply:
> I'm not too familiar with what findCompletion does, but this looks as if you're trying to find any completion for the given command ID.

The `findCompletion` function just hits the `CommandCompletionService`. However, while addressing this comment I realized that the `from`/`offset` parameter is exclusive, so I had to split finding the completion and assertions on the boundaries. 


> You should restrict the filter to accepting completions (because only those trigger deduplication)

You're right, done.


> you should make sure that you stop looking for completions before the completion (otherwise you might find another later completion for the same command ID; probably not for the current test cases, but maybe in the future?).

That's also correct, it's not problematic for the current test cases. It was fairly easy to change, so done.